pkg/services/sqlstore/session/session.go

```diff
@@ -68,6 +74,34 @@ func (gs *SessionDB) WithTransaction(ctx context.Context, callback func(*Session
 	return tx.sqlxtx.Commit()
 }
 
+func (gs *SessionDB) WithTransactionContext(ctx context.Context, callback func(context.Context) error) error {
+	txCtx := ctx
+
+	// If another transaction is already open, we just use that one instead of nesting.
+	sessionTx, ok := txCtx.Value(ContextSessionTxKey{}).(*SessionTx)
+	if !ok {
+		tx, err := gs.Beginx()
+		if err != nil {
+			return err
+		}
+
+		sessionTx = tx
+
+		// Save it in the context so it can be used by `SqlxSessionFromContext`.
+		txCtx = context.WithValue(ctx, ContextSessionTxKey{}, sessionTx)
+	}
+
+	if err := callback(txCtx); err != nil {
+		if rbErr := sessionTx.sqlxtx.Rollback(); rbErr != nil {
+			return fmt.Errorf("tx err: %v, rb err: %v", err, rbErr)
```

Comment:
```suggestion
			return fmt.Errorf("tx err: %w, rb err: %w", err, rbErr)
```

let's return both as errors, or at least `err`

Reply:
i will not change the current session, let me clean the pr up and will request a proper review -- and will change the errors!